src/domain/discounts/discount-form/form/discount-form-context.tsx

```diff
@@ -0,0 +1,288 @@
+import React, { useEffect, useState } from "react"
+import { FormProvider, useForm, useFormContext } from "react-hook-form"
+import { Option } from "../../../../types/shared"
+import { DiscountFormValues } from "./mappers"
+
+const defaultDiscount: DiscountFormValues = {
+  code: "",
+  type: "percentage",
+  value: undefined,
+  description: "",
+  allocation: "total",
+  valid_for: null,
+  is_dynamic: false,
+  regions: null,
+  starts_at: new Date(),
+  ends_at: null,
+}
+
+type DiscountFormProviderProps = {
+  discount?: DiscountFormValues
+  isEdit?: boolean
+  children?: React.ReactNode
+}
+
+export const DiscountFormProvider = ({
+  discount = defaultDiscount,
+  isEdit = false,
+  children,
+}: DiscountFormProviderProps) => {
+  const [regionsDisabled, setRegionsDisabled] = useState(false)
+  const [appliesToAll, setAppliesToAll] = useState(true)
+  const [isFreeShipping, setIsFreeShipping] = useState(false)
+  const [hasExpiryDate, setHasExpiryDate] = useState(false)
+  const [hasStartDate, setHasStartDate] = useState(false)
+  const [prevType, setPrevType] = useState<string | undefined>(undefined)
+  const [prevUsageLimit, setPrevUsageLimit] = useState<string>("")
+  const [prevValidDuration, setPrevValidDuration] = useState<string>("")
+  const [prevAllocation, setPrevAllocation] = useState<string | undefined>(
+    "total"
+  )
+  const [prevExpiryDate, setPrevExpiryDate] = useState<Date | undefined>(
+    undefined
+  )
+  const [startsAt, setStartsAt] = useState(discount.starts_at)
+  const [endsAt, setEndsAt] = useState(discount.ends_at)
+
+  const methods = useForm({ defaultValues: discount, reValidateMode: "onBlur" })
+
+  methods.register({
+    name: "allocation",
+    value: discount?.allocation || "total",
+  })
+
+  const setAllocation = (value) => {
+    methods.setValue("allocation", value)
+    setPrevAllocation(value)
+  }
+
+  const setConditionType = (value) => methods.setValue("condition_type", value)
```

Comment:
For the discount form we should add some way to manage multiple conditions and condition types in the hook. I’m a bit on the fence regarding how, but I would love to hear your thoughts. 

We could create individual properties in the form for each condition entity (products, types, tags, etc.). This would make it easy to control from a modal perspective in terms of the tables and adding/removing condition entities. However, it would bloat the form quite a bit since the fields would need to be controlled in some way. When we then submit the form we would map the array to their id’s and send them along to the api. 

A different option that is more closely related to how the api accepts the discount conditions on create and update would be to store an array of conditions. This would require us to loop the array (which has a max length of 5) when adding or removing condition entities from the discount, however when committing the changes to the api the conditions would require less mapping and wrangling in order to fit the api spec. This solution also bloats the form less since it’s a single field and not 5.

wdyt?

Reply:
I think I lean more towards an implementation of option 2. To give a bit more detail, I think that we should have a `conditions` array or object in the discount form, then leave each individual condition modal (i.e: product conditions modal, customer groups conditions modal, etc.) the freedom to implement/manage its own form. In other words, each conditions modal will be responsible of getting all the selected choices from the table or whatever, then it should call some method like `addCondition` that is exposed from the `useDiscountForm` hook to "register" the user input conditions.

Does this align with what you had in mind @pKorsholm 😄 ?

Reply:
So to summarise (and I agree with you guys) the idea we are going for is to have multiple (sub)forms i.e. we would have reference to data in the root form but delegate logic as much as possible to a nested modal/form.

Reply:
Yeah, the individual tables should definitely be responsible for managing their own conditions and adding/removing the entities correctly. @kasperkristensen brought up a good point that if we want to re-render the form a change in a child of the array will not trigger a re-render, then we would need to add an appropriate handler. This might become an issue when we get to showing the list of conditions that apply to a promotion in the modal before we save and update the promotion and thus invalidate the query.

We could consider creating an object for the state where the values map directly to the possible conditions plus a title, i.e.: 
```
{
  products?: {products: {id: string, title: string}[]}

  product_collections?: {product_collections: {id: string, title: string}[]}

  product_types?: {product_types: {id: string, title: string}[]}

  product_tags?: {product_tags: {id: string, title: string}[]}

  customer_groups?: {customer_groups: {id: string, title: string}[]}
}
```

Though this also feels like a hack to get it to work. I think I'm leaning more towards solution 1 and then having the mapper or another utility hook to transform the input for the api.